[quantization] PackedSequence support for quantized LSTM

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29585 [quantization] PackedSequence support for quantized LSTM**

Fixes https://github.com/pytorch/pytorch/issues/27963

Differential Revision: [D18436569](https://our.internmc.facebook.com/intern/diff/D18436569)

Commit:
[quantization] PackedSequence support for quantized LSTM

[ghstack-poisoned]

jianyuh commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -1194,6 +1194,53 @@ std::tuple<Tensor, Tensor, Tensor> quantized_lstm(
+    lstm_packed_miopen_stub(data.type().device_type(), output, hy, cy, data, batch_sizes, hx,
+            _params, has_biases, num_layers, dropout_p, train, bidirectional);
+    return std::make_tuple(std::move(output), std::move(hy), std::move(cy));
+  }
```
Do we need to add the same handler in Line 1160?

jamesr66a replied:
Actually, not sure this should even be here. I just directly copy-pasted from the regular LSTM function

BTW do we actually support cudnn for these ops?

jianyuh replied:
Appears to be yes: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/miopen/RNN_miopen.cpp#L889

jamesr66a replied:
@jianyuh I don't think those implementations (miopen and cudnn) actually support dynamic quantization

jianyuh replied:
Oh I thought you were asking if `lstm_packed_miopen` has cuda/cudnn implementation. Yep, the dynamic quantization is only supported on the CPU side.

jianyuh reviewed (approved):
LGTM! Thanks for finishing this so quickly!

jianyuh reviewed (dismissed):
Check the BC issue:
```
Nov 11 23:15:22 processing existing schema:  aten::quantized_lstm(Tensor input, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first, *, int? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor)
Nov 11 23:15:22 Can NOT find backward compatible schemas after changes for schema aten::quantized_lstm(Tensor input, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first, *, int? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor) from the following candidates:
Nov 11 23:15:22 [
Nov 11 23:15:22 aten::quantized_lstm.data(Tensor data, Tensor batch_sizes, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, *, int? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor)
Nov 11 23:15:22 	aten::quantized_lstm.tensor(Tensor input, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first, *, int? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor)
Nov 11 23:15:22 ]
Nov 11 23:15:22 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.
```

jamesr66a commented:
@jianyuh I believe the BC issue is a false positive: the only difference from before is it has the `.tensor` tag at the end. @houseroad do you mind taking a look?

houseroad commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3594,7 +3594,9 @@
 
 # Quantized RNN layers
-- func: quantized_lstm(Tensor input, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first, *, ScalarType? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor)
+- func: quantized_lstm.tensor(Tensor input, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first, *, ScalarType? dtype=None, bool use_dynamic=False) -> (Tensor, Tensor, Tensor)
```
Actually, this is a BC breaking change on mobile side (although no mobiles on mobile side yet).

The recommended change is keeping `quantized_lstm` whose overload name is empty, and adding `quantized_lstm.data`.

cc: @smessmer

Commit:
Update on "[quantization] PackedSequence support for quantized LSTM"


Fixes https://github.com/pytorch/pytorch/issues/27963

Differential Revision: [D18436569](https://our.internmc.facebook.com/intern/diff/D18436569)

[ghstack-poisoned]

Commit:
Update on "[quantization] PackedSequence support for quantized LSTM"


Fixes https://github.com/pytorch/pytorch/issues/27963

Differential Revision: [D18436569](https://our.internmc.facebook.com/intern/diff/D18436569)

[ghstack-poisoned]

Commit:
Update on "[quantization] PackedSequence support for quantized LSTM"


Fixes https://github.com/pytorch/pytorch/issues/27963

Differential Revision: [D18436569](https://our.internmc.facebook.com/intern/diff/D18436569)

[ghstack-poisoned]
